larpandoracontent/LArThreeDReco/LArTwoViewMatching/TwoViewThreeDKinkTool.cc

```diff
@@ -0,0 +1,480 @@
+{
+    for (IteratorList::const_iterator iIter1 = iteratorList.begin(), iIter1End = iteratorList.end(); iIter1 != iIter1End; ++iIter1)
+    {
+        for (IteratorList::const_iterator iIter2 = iIter1; iIter2 != iIter1End; ++iIter2)
+        {
+            if (iIter1 == iIter2)
+                continue;
+
+            try
+            {
+                const unsigned int nMatchedReUpsampledSamplingPoints1((*iIter1)->GetOverlapResult().GetNMatchedReUpsampledSamplingPoints());
+                const unsigned int nMatchedReUpsampledSamplingPoints2((*iIter2)->GetOverlapResult().GetNMatchedReUpsampledSamplingPoints());
+                IteratorList::const_iterator iIterA((nMatchedReUpsampledSamplingPoints1 >= nMatchedReUpsampledSamplingPoints2) ? iIter1 : iIter2);
+                IteratorList::const_iterator iIterB((nMatchedReUpsampledSamplingPoints1 >= nMatchedReUpsampledSamplingPoints2) ? iIter2 : iIter1);
+
+                Particle particle(*(*iIterA), *(*iIterB));
+                const LArPointingCluster pointingClusterA(pAlgorithm->GetCachedSlidingFitResult(particle.m_pClusterA));
+                const LArPointingCluster pointingClusterB(pAlgorithm->GetCachedSlidingFitResult(particle.m_pClusterB));
+
+                LArPointingCluster::Vertex vertexA, vertexB;
+                LArPointingClusterHelper::GetClosestVertices(pointingClusterA, pointingClusterB, vertexA, vertexB);
+
+                float transverseAB(std::numeric_limits<float>::max()), transverseBA(std::numeric_limits<float>::max());
+                float longitudinalAB(-std::numeric_limits<float>::max()), longitudinalBA(-std::numeric_limits<float>::max());
+
+                LArPointingClusterHelper::GetImpactParameters(vertexA, vertexB, longitudinalAB, transverseAB);
+                LArPointingClusterHelper::GetImpactParameters(vertexB, vertexA, longitudinalBA, transverseBA);
+
+                if (std::min(longitudinalAB, longitudinalBA) < m_minLongitudinalImpactParameter)
+                    continue;
+
+                if (std::min(transverseAB, transverseBA) > m_maxTransverseImpactParameter)
+                    continue;
+
+                const float cosTheta(-vertexA.GetDirection().GetCosOpeningAngle(vertexB.GetDirection()));
+
+                if (cosTheta < m_minImpactParameterCosTheta)
+                    continue;
+
+                const bool isALowestInX(this->IsALowestInX(pointingClusterA, pointingClusterB));
+                const CartesianVector splitPosition((vertexA.GetPosition() + vertexB.GetPosition()) * 0.5f);
+                const bool isThreeDKink(this->IsThreeDKink(pAlgorithm, particle, splitPosition, isALowestInX));
+
+                if (isThreeDKink != m_splitMode)
+                    continue;
+
+                // Construct the modification object
+                Modification modification;
+
+                if (m_splitMode)
```

Comment:
Biggest change: would you agree with removing the split mode entirely, and replacing the `if (m_splitMode)` with `if (isThreeDKink) {...} else {...}`. In that way, the 3D kink finding always drives whether we split (if it's kinked) or merge (if it's not). Can you test that? Thanks!

Reply:
Happy to test the version you suggest! Is it okay to just throw an exception in case the following check fails, in that case?
https://github.com/PandoraPFA/LArContent/pull/167/files#diff-fd01ba767797964c499ddec52b36d00b1251aee746f209ab4e217ced8d215a0eR392

Reply:
Hi, I think the options are pick a default operation (merge or split) in the case where the kink finding algorithm doesn't feel it can provide a decision, or make it skip over this modification entirely. This could be by raising an exception at that point, but I think you'll want to catch and continue back in the original calling function, rather than risk bailing the whole processing of this algorithm for this event. Picking a default split/merge doesn't sound so nice, but at least it means the tool will have responded to a problem identified in the matrix. It will have resolved an ambiguity, but perhaps not in the optimal way. Suggest you have a play about and see what you think is best. Thanks!

Reply:
Okay, great, thanks!